lib/functions/installs.bash

```diff
@@ -117,30 +117,199 @@ install_local_tool_versions() {
-      fi
-    done
+    plugins_installed=$(printf "%s" "$plugins_installed" | tr "\n" " " | awk '{$1=$1};1')
+    display_debug "install_local_tool_versions: plugins_installed='$plugins_installed'"
+    tools_installed=$(install_directory_tools_recursive "$search_path" "$plugins_installed")
   fi
 
-  if [ -z "$some_tools_installed" ]; then
+  if [ -z "$tools_installed" ]; then
     printf "Either specify a tool & version in the command\n"
     printf "OR add .tool-versions file in this directory\n"
     printf "or in a parent directory\n"
     exit 1
   fi
 }
 
+install_directory_tools_recursive() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=""
+
+  display_debug "install_directory_tools_recursive '$search_path': entered with plugins_installed='$plugins_installed'"
+
+  while [ "$search_path" != "/" ]; do
+    # install tools from files in current directory
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    tools_installed=$(install_directory_tools "$search_path" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$search_path': install_directory_tools returned tools_installed='$tools_installed'"
+
+    # terminate if all tools are installed
+    if [[ -n $(stringlist_a_subset_of_b "$plugins_installed" "$tools_installed") ]]; then
+      display_debug "install_directory_tools_recursive '$search_path': exiting because all known tools ($plugins_installed) are installed ($tools_installed)"
+      printf "%s\n" "$tools_installed"
+      return 0
+    fi
+
+    # go up a directory
+    search_path=$(dirname "$search_path")
+  done
+
+  # still got some tools to install...
+  # lets see if $ASDF_DEFAULT_TOOL_VERSIONS_FILENAME as actually
+  # an absolute path
+  if [ -f "$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME" ]; then
+    display_debug "--------------------------------------------------------------------------------------------------------------"
+    display_debug "attempting to treat \$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME as an absolute path: $ASDF_DEFAULT_TOOL_VERSIONS_FILENAME"
+    tools_installed=$(_install_directory_tools "" "$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME" "$plugins_installed" "$tools_installed")
+    display_debug "install_directory_tools_recursive '$ASDF_DEFAULT_TOOL_VERSIONS_FILENAME': install_directory_tools returned tools_installed='$tools_installed'"
+  fi
+
+  printf "%s\n" "$tools_installed"
+}
+
+install_directory_tools() {
+  local search_path=$1
+  local plugins_installed=$2
+  local tools_installed=$3
+  display_debug "install_directory_tools '$search_path': starting install. tools_installed='$tools_installed'"
+
+  # install tools from .tool-versions
+  # install order is the order listed in .tool-versions
+  file_name=$(asdf_tool_versions_filename)
+  tools_installed=$(_install_directory_tools "$search_path" "$file_name" "$plugins_installed" "$tools_installed")
```

Comment:
I'm used to wrapping everything in quotes. Is there a reason the `$()` on these two lines aren't quoted? Like this:

```
file_name="$(asdf_tool_versions_filename)"
```

Reply:
No reason. Will wrap it in quotes no problem

Reply:
Not necessary. Word splitting is not done on values in assignments (see [Bash Shell Parameters](https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameters)), i.e., quotes are not needed when assigning command substitutions (`$(...)`) or variables (`$var`).  E.g.:

```
foo='a space-filled value'
bar=$foo
echo $bar
```

> a space-filled value

```
foo=$(echo this is a command substitution with spaces)
echo $foo
```
> this is a command substitution with spaces